src/aws_encryption_sdk/materials_managers/__init__.py

```diff
@@ -40,38 +48,250 @@ class EncryptionMaterialsRequest(object):
+        :raises InvalidDataKeyError: if data key length does not match algorithm suite
+        """
+        self._validate_data_encryption_key(
+            data_encryption_key=data_encryption_key, keyring_trace=keyring_trace, required_flags=required_flags
+        )
+
+        data_key = _data_key_to_raw_data_key(data_key=data_encryption_key)
+
+        super(CryptographicMaterials, self).__setattr__("data_encryption_key", data_key)
+        self._keyring_trace.append(keyring_trace)
+
+    @property
+    def keyring_trace(self):
+        """Return a read-only version of the keyring trace.
+
+        :rtype: tuple
+        """
+        return tuple(self._keyring_trace)
+
+
+@attr.s(hash=False, init=False)
+class EncryptionMaterials(CryptographicMaterials):
     """Encryption materials returned by a crypto material manager's `get_encryption_materials` method.
 
     .. versionadded:: 1.3.0
 
-    :param algorithm: Algorithm to use for encrypting message
-    :type algorithm: aws_encryption_sdk.identifiers.Algorithm
-    :param data_encryption_key: Plaintext data key to use for encrypting message
-    :type data_encryption_key: aws_encryption_sdk.structures.DataKey
-    :param encrypted_data_keys: List of encrypted data keys
-    :type encrypted_data_keys: list of `aws_encryption_sdk.structures.EncryptedDataKey`
+    .. versionadded:: 1.5.0
+
+        The **keyring_trace** parameter.
+
+    .. versionadded:: 1.5.0
+
+        Most parameters are now optional.
+
+    :param Algorithm algorithm: Algorithm to use for encrypting message
+    :param DataKey data_encryption_key: Plaintext data key to use for encrypting message (optional)
+    :param encrypted_data_keys: List of encrypted data keys (optional)
+    :type encrypted_data_keys: list of :class:`EncryptedDataKey`
     :param dict encryption_context: Encryption context tied to `encrypted_data_keys`
-    :param bytes signing_key: Encoded signing key
+    :param bytes signing_key: Encoded signing key (optional)
+    :param keyring_trace: Any KeyRing trace entries (optional)
+    :type keyring_trace: list of :class:`KeyringTrace`
     """
 
-    algorithm = attr.ib(validator=attr.validators.instance_of(Algorithm))
-    data_encryption_key = attr.ib(validator=attr.validators.instance_of(DataKey))
-    encrypted_data_keys = attr.ib(validator=attr.validators.instance_of(set))
-    encryption_context = attr.ib(validator=attr.validators.instance_of(dict))
-    signing_key = attr.ib(default=None, validator=attr.validators.optional(attr.validators.instance_of(bytes)))
+    _encrypted_data_keys = attr.ib(
+        default=attr.Factory(list), validator=optional(deep_iterable(member_validator=instance_of(EncryptedDataKey)))
+    )
+    signing_key = attr.ib(default=None, repr=False, validator=optional(instance_of(bytes)))
+
+    def __init__(
+        self,
+        algorithm=None,
+        data_encryption_key=None,
+        encrypted_data_keys=None,
+        encryption_context=None,
```

Comment:
I chose not to include the Encryption Context in the material.  I did this because the top level caller _must_ have the encryption context.  I'm not saying we should remove it, just that it is slightly different.

Reply:
IMO we should include the encryption context so that it is more directly bound.

Specifically, the encryption context that needs to be used as part of these encryption materials, as possibly distinct from the encryption context that was written in the encrypted message.

Reply:
>encryption context that needs to be used as part of these encryption materials
That is the part that convinces me :)

I can get down with this.  The part I don't like is that the material is created by the CMM, when is the encryption context set?  The _best_ time would be when the material is created.  Otherwise I have to work to distinguish between "no encryption context" and "encryption context has not yet been set so these material are not yet valid"

Reply:
>  The part I don't like is that the material is created by the CMM

This is explicitly part of the CMM contract: any CMM MAY modify the encryption context in either direction.

> when is the encryption context set?

On initial creation of materials. The keyring MUST not modify the encryption context.

Reply:
Good, I'm cool with "initial creation of materials".  But once a CMM creates materials, downstream CMM's can ~not change the Encryption Context...  Or they have to re-create materials and clone things over.

i.e. this make is tedious to send an encryption context to the KMS keyring, but *not* return that encryption context to be serialized into the message format.

Reply:
>  Or they have to re-create materials and clone things over.

This. CMMs MAY modify the encryption context. They also MAY return whatever encryption/decryption materials that they deem appropriate. That MAY include a CMM layer that creates new materials based in part, whole, or not at all, on the materials returned from an underlying CMM.